packages/eslint-plugin/src/rules/unit-naming-rule.ts

```diff
@@ -1,265 +1,228 @@
-          isPrivate,
-          domain,
-        })
-        return fixer.replaceText(nameArg, x)
-      },
-    })
-    return
-  }
-
-  if ((imports.get(unitBody.callee.name) ?? unitBody.callee.name) === 'atom') {
-    if (!unitIdentifier.name.startsWith(atomPrefix)) {
-      context.report({
-        node: nameArg,
-        messageId: 'prefixMissing',
-        fix: (fixer) => {
-          const x = insertUnitName({
-            selfName: atomPrefix + unitIdentifier.name,
-            isPrivate,
-            domain,
-          })
-          return [
-            fixer.replaceText(nameArg, x),
-            fixer.replaceText(unitIdentifier, atomPrefix + unitIdentifier.name),
-          ]
-        },
-      })
-      return
-    }
-    if (!unitIdentifier.name.endsWith(atomPostfix)) {
-      context.report({
-        node: nameArg,
-        messageId: 'postfixMissing',
-        fix: (fixer) => {
-          const x = insertUnitName({
-            selfName: unitIdentifier.name + atomPostfix,
-            isPrivate,
-            domain,
-          })
-          return [
-            fixer.replaceText(nameArg, x),
-            fixer.replaceText(
-              unitIdentifier,
-              unitIdentifier.name + atomPostfix,
-            ),
-          ]
-        },
-      })
-      return
-    }
-  }
-}
-
-const insertUnitName = ({
-  selfName,
-  domain,
-  isPrivate = false,
-}: {
-  selfName: string
-  domain: AutoDomain | string | null
-  isPrivate?: boolean
-}) => {
-  if (isPrivate) selfName = '_' + selfName
-  if (domain === AutoDomain) return '`${name}.' + selfName + '`'
-  if (typeof domain === 'string') return `'${domain}.${selfName}'`
-  return `'${selfName}'`
-}
+import * as estree from 'estree'
+import { Rule } from 'eslint'
+import { patternNames, reatomFactoryPattern } from '../shared'
+
+type Domain = { is: 'absent' } | { is: 'static'; name: string } | { is: 'dynamic'; vary: string }
+
+interface Name {
+  domain: Domain
+  local: boolean
+  object: string | null
+  unit: string
+}
+
+export const unitNamingRule: Rule.RuleModule = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      recommended: true,
+      description: 'Ensures that all Reatom entities are correctly named.',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          atomPrefix: {
+            type: 'string',
+          },
+          atomPostfix: {
+            type: 'string',
+          },
+          domainVariable: {
+            type: 'string',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+    fixable: 'code',
+  },
+
+  create(context) {
+    const {
+      atomPrefix = '',
+      atomPostfix = '',
+      domainVariable = 'name',
+    }: {
+      atomPrefix?: string
+      atomPostfix?: string
+      domainVariable?: string
+    } = context.options[0] ?? {}
+
+    const domainScopes = [] as (Domain | null)[]
+    const idScopes = [] as (estree.Identifier | null)[]

```

Comment:
```suggestion
    const domainScopes: (Domain | null)[] = [] 
    const idScopes: (estree.Identifier | null)[] = []
```

Lets remove assertions.

Reply:
Second question is why do we need null? With following diff I have no errors in tests.

```diff
diff --git a/packages/eslint-plugin/src/rules/unit-naming-rule.ts b/packages/eslint-plugin/src/rules/unit-naming-rule.ts
index 5b5585db..5d906fa8 100644
--- a/packages/eslint-plugin/src/rules/unit-naming-rule.ts
+++ b/packages/eslint-plugin/src/rules/unit-naming-rule.ts
@@ -49,13 +49,15 @@ export const unitNamingRule: Rule.RuleModule = {
       domainVariable?: string
     } = context.options[0] ?? {}
 
-    const domainScopes = [] as (Domain | null)[]
-    const idScopes = [] as (estree.Identifier | null)[]
+    const domainScopes: Domain[] = []
+    const idScopes: estree.Identifier[] = []
 
     return {
       ':matches(VariableDeclarator, Property)'(node: estree.VariableDeclarator | estree.Property) {
         const id = node.type === 'Property' ? node.key : node.id
-        idScopes.push(id.type === 'Identifier' ? id : null)
+        if (id.type === 'Identifier') {
+          idScopes.push(id)
+        }
       },
       ':matches(VariableDeclarator, Property):exit'() {
         idScopes.pop()
@@ -73,7 +75,7 @@ export const unitNamingRule: Rule.RuleModule = {
         if (declaresDomain) domainScopes.push({ is: 'dynamic', vary: domainVariable })
         else if ('id' in node && node.id && node.id.name.startsWith('reatom')) {
           domainScopes.push({ is: 'static', name: node.id.name.replace('reatom', '') })
-        } else domainScopes.push(null)
+        }
       },
       [`:function:exit`](node: estree.Function) {
         domainScopes.pop()

```

Reply:
1. there are no problems with assertions when they are used to specify the element type of an empty array literal
2. `*:exit` listeners unconditionally pop elements from `domainScopes` / `idScopes` stacks even if their corresponding enter listeners did not add a domain/identifier scope. although tests do not capture such cases, removing null stack elements will cause any non-factory functions, destructured variables and computed-keyed properties to break the rule